File: NewsApp/app/src/main/java/com/example/android/topic41/data/repository/ArticlesRepository.java
package com.example.android.topic41.data.repository;

import com.example.android.topic41.data.database.CacheInterface;
import com.example.android.topic41.data.network.NetworkInterface;
import com.example.android.topic41.domain.util.Article;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.Single;

public class ArticlesRepository implements ArticlesRepositoryInterface {
    private NetworkInterface network;
    private CacheInterface cache;
    private Map<String, Long> cacheTime = new HashMap<>();

    private static final int TIME_TO_REFRESH = 60_000;


    public ArticlesRepository(CacheInterface cache, NetworkInterface network) {
        this.cache = cache;
        this.network = network;
    }


    @Override
    public Single<List<Article>> loadArticles(String theme) {
        long currentTime = System.currentTimeMillis();
        boolean isCacheFull = cacheTime.get(theme) != null && currentTime - cacheTime.get(theme) < TIME_TO_REFRESH;
        if (isCacheFull) {
            return cache.getArticlesByTheme(theme);
        } else {
            return network.createObservable(theme)
                    .map(articles -> prepareForCache(theme, articles))
                    .map(this::refreshCache);
        }
    }


    public List<Article> prepareForCache(String theme, List<Article> articles) {
        long currentTime = System.currentTimeMillis();
        cacheTime.put(theme, currentTime);
        for (Article article : articles) {
            article.setTime(currentTime);
            article.setTheme(theme);
        }
        return articles;
    }

    public List<Article> refreshCache(List<Article> articles) {
        if (articles.size() > 0) {
            cache.refresh(articles, articles.get(0).getTime());
        }
        return articles;
    }
}

File: NewsApp/app/src/main/java/com/example/android/topic41/data/repository/ArticlesRepository.kt
package com.example.android.topic41.data.repository

import com.example.android.topic41.data.database.CacheInterface
import com.example.android.topic41.data.network.NetworkInterface
import com.example.android.topic41.domain.util.Article
import io.reactivex.Single
import java.util.*


class ArticlesRepository(private val cache: CacheInterface,
                         private val network: NetworkInterface) : ArticlesRepositoryInterface {

    private val cacheTime: MutableMap<String, Long> = HashMap()

    override fun loadArticles(theme: String): Single<List<Article>> {
        val currentTime = System.currentTimeMillis()
        val isCacheFull = cacheTime[theme] != null && currentTime - cacheTime[theme]!! < TIME_TO_REFRESH
        return if (isCacheFull) {
            cache.getArticlesByTheme(theme)
        } else {
            network.createObservable(theme)
                    .map { articles: List<Article> -> prepareForCache(theme, articles) }
                    .map { articles: List<Article> -> refreshCache(articles) }
        }
    }

    private fun prepareForCache(theme: String, articles: List<Article>): List<Article> {
        val currentTime = System.currentTimeMillis()
        cacheTime[theme] = currentTime
        return articles
                .map { it.time = currentTime; it.theme = theme; it }
    }

    private fun refreshCache(articles: List<Article>): List<Article> {
        if (articles.isNotEmpty()) {
            cache.refresh(articles, articles[0].time)
        }
        return articles
    }

    companion object {
        private const val TIME_TO_REFRESH = 60_000
    }
}
